### PR TITLE
31 use w projection in simple causal models

### LINKED ISSUE #31: Use W-Projection in Simple Causal Models
- [x] Implement W-Projection as defined by Hopkins.
- [x] Update tests as required.

At this state, if a given variable is included in _W_ but not in the W-projection, an exception will be raised.

### PR DESCRIPTION
**Pull request by mahykari**

### CODE REVIEW

File: estrest/src/causality/causal_model.py [hunk 1]
```diff
@@ -86,3 +86,42 @@ def evaluate(self):
...
+            vars_.update(self.deps[v])
+        return vars_
+
+    # W-projection is achieved with the following:
+    # * Updating deps to remove any v not in projection variables
+    #   from deps[u] for all u.
+    # * Fixing any variable v in projection variables where deps[v] is empty
+    #   (v only depended on variables not in the projection.)
+    # * Removing variables not in the projection
+    def w_projection(
```
Comment (seyhani) [inline estrest/src/causality/causal_model.py]:
```suggestion
    def get_w_projection(
```

File: estrest/src/causality/causal_model.py [hunk 2]
```diff
@@ -86,3 +86,42 @@ def evaluate(self):
 
     def get_var_names(self):
         return self.fns.keys()
+
+    def _w_projection_vars(self, X: str, Y: str) -> Set[str]:
```
Comment (seyhani) [inline estrest/src/causality/causal_model.py]:
```suggestion
    def __get_w_projection_vars(self, X: str, Y: str) -> Set[str]:
```

File: estrest/src/causality/causal_model.py [hunk 3]
```diff
@@ -86,3 +86,42 @@ def evaluate(self):
...
+        for var in model.deps:
+            deps = model.deps[var]
+            model.deps[var] = list(filter(lambda v: v in pv, deps))
+
+        for var in pv:
+            if len(model.deps[var]) > 0:
+                continue
+            model.fns[var] = (
+                lambda val_=model.vals[var]: lambda vals: val_
+            )()
```
Comment (seyhani) [inline estrest/src/causality/causal_model.py]:
```suggestion
            if len(model.deps[var]) == 0:
                model.fns[var] = (
                    lambda val_=model.vals[var]: lambda vals: val_
                )()
```

File: estrest/test/test_causal_model.py [hunk 4]
```diff
@@ -32,9 +32,36 @@ def test_intervention(self):
...
+    def test_w_projection_vars(self):
+        m = CausalModel()
+        m.add('A', lambda vals: True, [])
+        m.add('B', lambda vals: False, [])
+        m.add('C', lambda vals: vals['A'] and vals['B'], ['A', 'B'])
+        m.add('D', lambda vals: vals['C'] and True, ['C'])
+
+        wp = m._w_projection_vars('A', 'D')
+        self.assertEqual(wp, {'A', 'B', 'C', 'D'})
+
```
Comment (seyhani) [inline estrest/test/test_causal_model.py]: Add another test to check whether some variables are being removed.

**Review by seyhani — CHANGES_REQUESTED** [review]

File: estrest/test/test_causal_model.py (hunk 4 above)
Comment (mahykari) [reply estrest/test/test_causal_model.py]: Done. I used the SUZY_BILLY model. For checking Billy not being a cause, ST must be removed.

**Review by seyhani — APPROVED** [review]

**Pull request merged**